Try to handle enums that fail isEnum

jsoref opened the pull request:
https://www.logicbig.com/how-to/code-snippets/jcode-reflection-field-isenumconstant.html

Current code, conceptually:
```Groovy
d=java.util.concurrent.TimeUnit.DAYS
[d.class.isEnum(),d.getDeclaringClass().class.isEnum()]

Result: [false, false]
```

Attempted conceptual supplement:
```Groovy
declaredFields = java.util.concurrent.TimeUnit.class.getDeclaredFields();
declaredFields.collect { [it.getName(), it.isEnumConstant()] }

Result: [[NANOSECONDS, true], [MICROSECONDS, true], [MILLISECONDS, true], [SECONDS, true], [MINUTES, true], [HOURS, true], [DAYS, true], [C0, false], [C1, false], [C2, false], [C3, false], [C4, false], [C5, false], [C6, false], [MAX, false], [$VALUES, false]]
```


See [JENKINS-52945](https://issues.jenkins-ci.org/browse/JENKINS-52945).

### Proposed changelog entries

* Relax JEP-200 to handle enums (like `java.util.concurrent.TimeUnit`) that fail `class.isEnum()`

### Submitter checklist

- [ ] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests

### Desired reviewers

@jglick 

Note:
* I haven't actually compiled this, and I'm not really a Java 8 person (more like java 5). I /think/ this might do what I want.
* This definitely should have tests (since it's relaxing JEP-200)
* This should have multiple reviewers once it actually vaguely works
* Assuming someone added a special whitelist for TimeUnit, that should be removed once this goes in

jsoref committed:
(DRAFT) Try to handle enums that fail isEnum

https://www.logicbig.com/how-to/code-snippets/jcode-reflection-field-isenumconstant.html

Current code, conceptually:
Groovy:
d=java.util.concurrent.TimeUnit.DAYS
[d.class.isEnum(),d.getDeclaringClass().class.isEnum()]

Result: [false, false]

Attempted conceptual supplement:
Groovy:
declaredFields = java.util.concurrent.TimeUnit.class.getDeclaredFields();
declaredFields.collect { [it.getName(), it.isEnumConstant()] }

Result: [[NANOSECONDS, true], [MICROSECONDS, true], [MILLISECONDS, true], [SECONDS, true], [MINUTES, true], [HOURS, true], [DAYS, true], [C0, false], [C1, false], [C2, false], [C3, false], [C4, false], [C5, false], [C6, false], [MAX, false], [$VALUES, false]]

This should address JENKINS-52945

oleg-nenashev reviewed (commented):
I may have no time this week, let's see

jsoref committed:
Try to properly roll together the various pieces.

Here's the Groovy I used for testing:
```Groovy
tu=java.util.concurrent.TimeUnit;
d=tu.MINUTES;
dc=d.declaringClass;
if (dc.isEnum()) {
  name=d.name();
  declaredFields=dc.declaredFields;
  v = tu.values();
  for (int i = 0; i < v.length; i++) {
    f = declaredFields[i];
    if(f.isEnumConstant() && v[i] == d) {
      println(v[i].name());
      println(f);
      println("permitting " + name + " since it is an enum");
    }
  }
}
```

jglick commented on core/src/main/java/jenkins/security/ClassFilterImpl.java:
```diff
@@ -155,10 +157,25 @@ public boolean isBlacklisted(Class _c) {
                 LOGGER.log(Level.FINE, "permitting {0} since it is a throwable", name);
                 return false;
             }
-            if (Enum.class.isAssignableFrom(c)) { // Class.isEnum seems to be false for, e.g., java.util.concurrent.TimeUnit$6
```
This comment describes exactly why I am using the construction above rather than `Class.isEnum`, for exactly the case of `TimeUnit` you bring up, so I do not understand what the purpose of this patch is.

jsoref commented:
It looks like jenkinsci/remoting#282 is the better fix for this...
I was grasping at straws...

jglick commented:
They are not really that closely related. This PR is touching JEP-200 code which would accept or reject (de)serialization of a class for security reasons; that PR is touching code which _warns_ about (de)serialization of anonymous classes for stability / maintainability reasons.
